File: backend/src/Modules/ZipZap.Modules.Catalog/Domain/Store.cs
using ZipZap.BuildingBlocks.Domain;

namespace ZipZap.Modules.Catalog.Domain;

/// <summary>Status operacyjny sklepu (czy przyjmuje zamówienia).</summary>
public enum StoreStatus { Open, Closed, TemporarilyUnavailable }

/// <summary>Sklep — korzeń najemcy (StoreId = <see cref="Entity.Id"/>).</summary>
public sealed class Store : AggregateRoot
{
    public string Name { get; private set; } = default!;
    public string Slug { get; private set; } = default!;
    public string? Description { get; private set; }
    public string City { get; private set; } = default!;
    public string? Address { get; private set; }
    public string? Phone { get; private set; }

    /// <summary>Logo sklepu (URL) — branding widoczny dla klienta. Puste = brak.</summary>
    public string? LogoUrl { get; private set; }

    /// <summary>Współrzędne sklepu — do proponowania sklepów wg odległości. Null = nieustalone.</summary>
    public double? Latitude { get; private set; }
    public double? Longitude { get; private set; }

    /// <summary>Prowizja ZipZap od wartości koszyka (0.10 = 10%).</summary>
    public decimal CommissionRate { get; private set; }

    /// <summary>Minimalna wartość koszyka (produktów), by złożyć zamówienie.</summary>
    public decimal MinimumOrderValue { get; private set; }

    /// <summary>Administracyjne włączenie sklepu (widoczność na platformie).</summary>
    public bool IsActive { get; private set; }

    /// <summary>Status operacyjny (przyjmowanie zamówień).</summary>
    public StoreStatus Status { get; private set; }

    public DateTime CreatedAtUtc { get; private set; }

    /// <summary>Sklep przyjmuje zamówienia tylko gdy aktywny i otwarty.</summary>
    public bool IsAcceptingOrders => IsActive && Status == StoreStatus.Open;

    private Store() { } // EF

    private Store(Guid id, string name, string slug, string? description,
        string city, string? address, string? phone, decimal commissionRate, decimal minimumOrderValue)
    {
        Id = id;
        Name = name;
        Slug = slug;
        Description = description;
        City = city;
        Address = address;
        Phone = phone;
        CommissionRate = commissionRate;
        MinimumOrderValue = minimumOrderValue;
        IsActive = true;
        Status = StoreStatus.Open;
        CreatedAtUtc = DateTime.UtcNow;
    }

    public static Store Create(string name, string slug, string? description,
        string city, string? address, string? phone, decimal commissionRate, decimal minimumOrderValue)
        => new(Guid.NewGuid(), name.Trim(), slug, description, city.Trim(), address, phone, commissionRate, minimumOrderValue);

    public void UpdateCommissionRate(decimal rate) => CommissionRate = rate;
    public void UpdateMinimumOrderValue(decimal value) => MinimumOrderValue = value;
    public void SetStatus(StoreStatus status) => Status = status;
    public void Activate() => IsActive = true;
    public void Deactivate() => IsActive = false;

    public void SetLogoUrl(string? logoUrl)
        => LogoUrl = string.IsNullOrWhiteSpace(logoUrl) ? null : logoUrl.Trim();

    /// <summary>Ustawia współrzędne; przekazanie null czyści (np. po zmianie adresu do ponownego geokodowania).</summary>
    public void SetLocation(double? latitude, double? longitude)
    {
        if (latitude is < -90 or > 90) throw new ArgumentOutOfRangeException(nameof(latitude), "Szerokość musi być w zakresie -90..90.");
        if (longitude is < -180 or > 180) throw new ArgumentOutOfRangeException(nameof(longitude), "Długość musi być w zakresie -180..180.");
        Latitude = latitude;
        Longitude = longitude;
    }
}

File: mobile/lib/models/store.dart
class Store {
  final String id;
  final String name;
  final String slug;
  final String city;
  final String status; // Open | Closed | TemporarilyUnavailable
  final String? description;
  final String? address;
  final String? phone;
  final double minimumOrderValue;
  final bool isAcceptingOrders;
  final String? logoUrl;
  final double? latitude;
  final double? longitude;
  final double? distanceKm;

  Store({
    required this.id,
    required this.name,
    required this.slug,
    required this.city,
    required this.status,
    required this.description,
    required this.address,
    required this.phone,
    required this.minimumOrderValue,
    required this.isAcceptingOrders,
    this.logoUrl,
    this.latitude,
    this.longitude,
    this.distanceKm,
  });

  bool get hasLocation => latitude != null && longitude != null;

  /// Odległość do sklepu w formie „2,3 km" / „850 m" (null gdy nieznana).
  String? get distanceLabel {
    final d = distanceKm;
    if (d == null) return null;
    if (d < 1) return '${(d * 1000).round()} m';
    return '${d.toStringAsFixed(1).replaceAll('.', ',')} km';
  }

  factory Store.fromJson(Map<String, dynamic> j) => Store(
        id: j['id'].toString(),
        name: j['name'] ?? '',
        slug: j['slug'] ?? '',
        city: j['city'] ?? '',
        status: j['status'] ?? 'Open',
        description: j['description'],
        address: j['address'],
        phone: j['phone'],
        minimumOrderValue: (j['minimumOrderValue'] as num?)?.toDouble() ?? 0,
        isAcceptingOrders: j['isAcceptingOrders'] == true,
        logoUrl: (j['logoUrl'] as String?)?.isNotEmpty == true ? j['logoUrl'] as String : null,
        latitude: (j['latitude'] as num?)?.toDouble(),
        longitude: (j['longitude'] as num?)?.toDouble(),
        distanceKm: (j['distanceKm'] as num?)?.toDouble(),
      );
}
